### ISSUE TITLE
Unexplained behavior of not finding violations

### ISSUE THREAD

**Issue by glensc** [phpmd/phpmd#1214]
- PHPMD version: 2.15.0
- PHP Version: 8.2.24
- Installation type: composer
- Operating System / Distribution & Version: macOS Sonoma 14.7 

### Current Behavior

I'm trying to use `DuplicatedArrayKey` rule, and it's not finding the violation.

### Expected Behavior

Report the DuplicatedArrayKey violation

### Steps To Reproduce:

```xml
# cat phpmd.xml
<?xml version="1.0"?>
<ruleset name="Custom Ruleset for DuplicatedArrayKey"
         xmlns="https://pmd.github.io/pmd-5.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="https://pmd.github.io/pmd-5.0.0 https://pmd.github.io/pmd-5.0.0.xsd"
         xsi:noNamespaceSchemaLocation="https://pmd.github.io/pmd-5.0.0.xsd">

    <description>
        This ruleset applies only the DuplicatedArrayKey rule.
    </description>

    <!-- Define the DuplicatedArrayKey rule -->
    <rule name="DuplicatedArrayKey"
          message="The array contains duplicated keys, which can lead to unexpected behavior."
          class="PHPMD\Rule\CleanCode\DuplicatedArrayKey">
        <description>
            This rule detects duplicated keys in arrays.
        </description>
    </rule>
</ruleset>
```

```php
# cat config/bp.php
<?php

return [
    'b2b_bank_name' => 'AS Swedbank',
    "b2b_bank_name" => 'AS Swedbank',
];
```

```
$ phpmd config/bp.php ansi phpmd.xml

Found 0 violations and 0 errors in 136ms

No mess detected
```

So, I downloaded test from PR implemented the change:
- https://github.com/phpmd/phpmd/pull/484

Pay attention that second run outcome is different even the files are not modified between the phpmd invocations

```
$ wget https://raw.githubusercontent.com/eeree/phpmd/a040a41c27c3841332f20e93ed2349192a6867d8/src/test/resources/files/Rule/CleanCode/DuplicatedArrayKey/testRuleAppliesCorrectlyToMultipleArrays.php

$ vendor/bin/phpmd testRuleAppliesCorrectlyToMultipleArrays.php ansi phpmd.xml -vvv
ResultCache is not enabled.

FILE: testRuleAppliesCorrectlyToMultipleArrays.php
-------------------------------------------------------------------------------------------------------
 20 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 29 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 29 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 37 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.

Found 4 violations and 0 errors in 125ms

$ cp config/bp.php testRuleAppliesCorrectlyToMultipleArrays.php
cp: overwrite 'testRuleAppliesCorrectlyToMultipleArrays.php'? y
$ vendor/bin/phpmd testRuleAppliesCorrectlyToMultipleArrays.php ansi phpmd.xml -vvv
ResultCache is not enabled.

FILE: testRuleAppliesCorrectlyToMultipleArrays.php
-------------------------------------------------------------------------------------------------------
 20 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 29 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 29 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.
 37 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.

Found 4 violations and 0 errors in 125ms
$ vendor/bin/phpmd testRuleAppliesCorrectlyToMultipleArrays.php ansi phpmd.xml -vvv
ResultCache is not enabled.

Found 0 violations and 0 errors in 91ms

No mess detected

$ cat .phpmd.result-cache.php
cat: .phpmd.result-cache.php: No such file or directory
```

what is happening?
1. why it doesn't detect violations from config/bp.php?
2. why overwriting the file with bp.php contents it first reports errors from previous file
3. why second run outcome is different

looks like something is cached? but where?

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by glensc** [comment]
while the cache-like issue can't be explained, I found if the code is wrapped in a function, then it will detect the error


```php
# vendor/bin/phpmd testRuleAppliesCorrectlyToMultipleArrays.php ansi phpmd.xml -vvv
ResultCache is not enabled.

FILE: testRuleAppliesCorrectlyToMultipleArrays.php
-------------------------------------------------------------------------------------------------------
 5 | VIOLATION | The array contains duplicated keys, which can lead to unexpected behavior.

Found 1 violation and 0 errors in 98ms
```

```php
# cat testRuleAppliesCorrectlyToMultipleArrays.php
<?php

function testRuleAppliesCorrectlyToMultipleArrays()
{
	return [
		'b2b_bank_name' => 'AS Swedbank',
		11=>1,
		"b2b_bank_name" => 'AS Swedbank',
	];
}
```

**Comment by AJenbo** [comment]
Duplicate of https://github.com/phpmd/phpmd/issues/681
